python/cuda_parallel/tests/test_reduce.py

```diff
@@ -75,11 +84,237 @@ def min_op(a, b):
     dtypes = [numpy.int32, numpy.int64]
     h_inits = [numpy.array([], dt) for dt in dtypes]
     h_inputs = [numpy.array([], dt) for dt in dtypes]
-    d_outputs = [cuda.device_array(1, dt) for dt in dtypes]
-    d_inputs = [cuda.to_device(h_inp) for h_inp in h_inputs]
+    d_outputs = [numba.cuda.device_array(1, dt) for dt in dtypes]
+    d_inputs = [numba.cuda.to_device(h_inp) for h_inp in h_inputs]
 
     reduce_into = cudax.reduce_into(d_inputs[0], d_outputs[0], min_op, h_inits[0])
 
     for ix in range(3):
         with pytest.raises(TypeError, match=r"^dtype mismatch: __init__=int32, __call__=int64$"):
-          reduce_into(None, d_inputs[int(ix == 0)], d_outputs[int(ix == 1)], h_inits[int(ix == 2)])
+          reduce_into(None, d_inputs[int(ix == 0)], d_outputs[int(ix == 1)], None, h_inits[int(ix == 2)])
+
+
+# fmt: on
+def _test_device_sum_with_iterator(
+    l_varr, start_sum_with, i_input, dtype_inp, dtype_out, use_numpy_array
+):
+    def add_op(a, b):
+        return a + b
+
+    expected_result = start_sum_with
+    for v in l_varr:
+        expected_result = add_op(expected_result, v)
+
+    if use_numpy_array:
+        h_input = numpy.array(l_varr, dtype_inp)
+        d_input = numba.cuda.to_device(h_input)
+    else:
+        d_input = i_input
+
+    d_output = numba.cuda.device_array(1, dtype_out)  # to store device sum
+
+    h_init = numpy.array([start_sum_with], dtype_out)
+
+    reduce_into = cudax.reduce_into(
+        d_in=d_input, d_out=d_output, op=add_op, init=h_init
+    )
+
+    temp_storage_size = reduce_into(
+        None, d_in=d_input, d_out=d_output, num_items=len(l_varr), init=h_init
+    )
+    d_temp_storage = numba.cuda.device_array(temp_storage_size, dtype=numpy.uint8)
+
+    reduce_into(d_temp_storage, d_input, d_output, len(l_varr), h_init)
+
+    h_output = d_output.copy_to_host()
+    assert h_output[0] == expected_result
+
+
+def mul2(val):
+    return 2 * val
+
+
+def mul3(val):
+    return 3 * val
+
+
+SUPPORTED_VALUE_TYPE_NAMES = (
```

Comment:
Why not just use `numpy` types, which are trivially convertible to numba types via `numba.from_dtype(...)`?